lib/zstream/encryption_coder/aes.ex

```diff
@@ -0,0 +1,196 @@
+      {<<>>, %{state | buffer: input}}
+    else
+      block_count = div(input_size, @aes_block_size)
+      blocks = Enum.map(counter..(counter + block_count - 1), &<<&1::unsigned-little-128>>)
+      plaintext_size = block_count * @aes_block_size
+      <<plaintext::binary-size(plaintext_size), new_buffer::binary>> = input
+      cipher = :crypto.exor(plaintext, :crypto.crypto_update(crypto_state, blocks))
+
+      {
+        cipher,
+        %{
+          state
+          | mac_state: :crypto.mac_update(state.mac_state, cipher),
+            counter: counter + block_count,
+            buffer: new_buffer
+        }
+      }
+    end
+  end
+
+  @impl true
+  def close(%State{buffer: buffer, counter: counter, crypto_state: crypto_state} = state) do
+    buffer_size = byte_size(buffer)
+
+    {final_encrypted, final_state} =
+      if buffer_size > 0 do
+        last_block = <<counter::unsigned-little-128>>
+
+        cipher =
+          :crypto.exor(
+            buffer,
+            binary_part(:crypto.crypto_update(crypto_state, last_block), 0, buffer_size)
+          )
+
+        final_state = %{
+          state
+          | mac_state: :crypto.mac_update(state.mac_state, cipher),
+            counter: counter + 1,
+            buffer: <<>>
+        }
+
+        {cipher, final_state}
+      else
+        {<<>>, state}
+      end
+
+    # https://www.winzip.com/win/en/aes_info.html#auth-faq
+    auth_code = binary_part(:crypto.mac_final(final_state.mac_state), 0, 10)
+    _should_be_empty = :crypto.crypto_final(final_state.crypto_state)
+
+    final_encrypted <> auth_code
+  end
+
+  # https://www.winzip.com/en/support/aes-encryption/#comp-method
+  @impl true
+  def general_purpose_flag do
+    # which means encrypted (0x0001)
+    0x0001
+  end
+
+  # https://www.winzip.com/en/support/aes-encryption/#comp-method
+  # a compression method of 99 is used to indicate the presence of an AES-encrypted file
+  @impl true
+  def compression_method, do: 99
+
+  # https://www.winzip.com/en/support/aes-encryption/#extra-data
+  @impl true
+  def extra_field_data(options) do
+    {coder, _options} = Keyword.fetch!(options, :coder)
+    {_encryption_coder, encryption_options} = Keyword.fetch!(options, :encryption_coder)
+    key_size = Keyword.get(encryption_options, :key_size, 256)
+    ae_version = Keyword.get(encryption_options, :ae_version, 2)
+
+    <<
+      # Extra field header ID
+      0x9901::little-size(16),
+      # Data size
+      7::little-size(16),
+      # Integer version number specific to the zip vendor, 0x0001 ae-1, 0x0002 ae-2
+      ae_version::little-size(16),
+      # 2-character vendor ID
+      "AE"::binary,
+      # Integer mode value indicating AES encryption strength
+      @aes_mode_indicators[key_size]::little-size(8),
+      # Actual compression method used (8=deflate, 0=stored)
+      coder.compression_method()::little-size(16)
+    >>
+  end
+
+  @impl true
+  def version_needed_to_extract, do: 51
```

Comment:
Can you point me to the source of this number? I can't find the actual number in the spec

> WinZip sets the version needed to extract and version made by fields in the local and central headers to the same values it would use if the file were not encrypted.